File: src/be/digitalia/fosdem/parsers/EventsParser.java
package be.digitalia.fosdem.parsers;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import org.xmlpull.v1.XmlPullParser;

import android.text.TextUtils;
import be.digitalia.fosdem.model.Day;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.Link;
import be.digitalia.fosdem.model.Person;
import be.digitalia.fosdem.model.Track;

/**
 * Main parser for FOSDEM schedule data in pentabarf XML format.
 * 
 * @author Christophe Beyls
 * 
 */
public class EventsParser extends IterableAbstractPullParser<Event> {

	private static final String BELGIUM_TIMEZONE = "GMT+1";
	private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd", Locale.US);

	{
		// All dates and times are in Belgium timezone
		DATE_FORMAT.setTimeZone(TimeZone.getTimeZone(BELGIUM_TIMEZONE));
	}

	// Calendar used to compute the events time, according to Belgium timezone
	private final Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone(BELGIUM_TIMEZONE), Locale.US);

	private Day currentDay;
	private String currentRoom;
	private Track currentTrack;
	// Recycled objects (improving parser performance)
	private List<Person> persons = new ArrayList<Person>();
	private List<Link> links = new ArrayList<Link>();

	/**
	 * Returns the hours portion of a time string in the "hh:mm" format, without allocating objects.
	 * 
	 * @param time
	 * @return hours
	 */
	private static int getHours(String time) {
		return (Character.getNumericValue(time.charAt(0)) * 10) + Character.getNumericValue(time.charAt(1));
	}

	/**
	 * Returns the minutes portion of a time string in the "hh:mm" format, without allocating objects.
	 * 
	 * @param time
	 * @return minutes
	 */
	private static int getMinutes(String time) {
		return (Character.getNumericValue(time.charAt(3)) * 10) + Character.getNumericValue(time.charAt(4));
	}

	@Override
	protected boolean parseHeader(XmlPullParser parser) throws Exception {
		while (!isEndDocument()) {
			if (isStartTag("schedule")) {
				return true;
			}

			parser.next();
		}
		return false;
	}

	@Override
	protected Event parseNext(XmlPullParser parser) throws Exception {
		while (!isNextEndTag("schedule")) {
			if (isStartTag()) {
				String name = parser.getName();

				if ("day".equals(name)) {
					currentDay = new Day();
					currentDay.setIndex(Integer.parseInt(parser.getAttributeValue(null, "index")));
					currentDay.setDate(DATE_FORMAT.parse(parser.getAttributeValue(null, "date")));
				} else if ("room".equals(name)) {
					currentRoom = parser.getAttributeValue(null, "name");
				} else if ("event".equals(name)) {
					Event event = new Event();
					event.setId(Integer.parseInt(parser.getAttributeValue(null, "id")));
					event.setDay(currentDay);
					event.setRoomName(currentRoom);
					event.setPersons(persons);
					event.setLinks(links);
					// reset persons and links list
					persons.clear();
					links.clear();
					String duration = null;
					String trackName = "";
					Track.Type trackType = Track.Type.other;

					while (!isNextEndTag("event")) {
						if (isStartTag()) {
							name = parser.getName();

							if ("start".equals(name)) {
								String time = parser.nextText();
								if (!TextUtils.isEmpty(time)) {
									calendar.setTime(currentDay.getDate());
									calendar.set(Calendar.HOUR_OF_DAY, getHours(time));
									calendar.set(Calendar.MINUTE, getMinutes(time));
									event.setStartTime(calendar.getTime());
								}
							} else if ("duration".equals(name)) {
								duration = parser.nextText();
							} else if ("slug".equals(name)) {
								event.setSlug(parser.nextText());
							} else if ("title".equals(name)) {
								event.setTitle(parser.nextText());
							} else if ("subtitle".equals(name)) {
								event.setSubTitle(parser.nextText());
							} else if ("track".equals(name)) {
								trackName = parser.nextText();
							} else if ("type".equals(name)) {
								try {
									trackType = Enum.valueOf(Track.Type.class, parser.nextText());
								} catch (Exception e) {
									// trackType will be "other"
								}
							} else if ("abstract".equals(name)) {
								event.setAbstractText(parser.nextText());
							} else if ("description".equals(name)) {
								event.setDescription(parser.nextText());
							} else if ("persons".equals(name)) {
								while (!isNextEndTag("persons")) {
									if (isStartTag("person")) {
										Person person = new Person();
										person.setId(Integer.parseInt(parser.getAttributeValue(null, "id")));
										person.setName(parser.nextText());

										persons.add(person);
									}
								}
							} else if ("links".equals(name)) {
								while (!isNextEndTag("links")) {
									if (isStartTag("link")) {
										Link link = new Link();
										link.setUrl(parser.getAttributeValue(null, "href"));
										link.setDescription(parser.nextText());

										links.add(link);
									}
								}
							} else {
								skipToEndTag();
							}
						}
					}

					if ((event.getStartTime() != null) && !TextUtils.isEmpty(duration)) {
						calendar.add(Calendar.HOUR_OF_DAY, getHours(duration));
						calendar.add(Calendar.MINUTE, getMinutes(duration));
						event.setEndTime(calendar.getTime());
					}

					if ((currentTrack == null) || !trackName.equals(currentTrack.getName()) || (trackType != currentTrack.getType())) {
						currentTrack = new Track(trackName, trackType);
					}
					event.setTrack(currentTrack);

					return event;
				} else {
					skipToEndTag();
				}
			}
		}
		return null;
	}
}

File: app/src/main/java/be/digitalia/fosdem/parsers/EventsParser.kt
package be.digitalia.fosdem.parsers

import be.digitalia.fosdem.model.Attachment
import be.digitalia.fosdem.model.Day
import be.digitalia.fosdem.model.DetailedEvent
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.model.EventDetails
import be.digitalia.fosdem.model.Link
import be.digitalia.fosdem.model.Person
import be.digitalia.fosdem.model.Track
import be.digitalia.fosdem.utils.isEndDocument
import be.digitalia.fosdem.utils.isNextEndTag
import be.digitalia.fosdem.utils.isStartTag
import be.digitalia.fosdem.utils.skipToEndTag
import be.digitalia.fosdem.utils.xmlPullParserFactory
import okio.BufferedSource
import org.xmlpull.v1.XmlPullParser
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import javax.inject.Inject
import javax.inject.Named

/**
 * Main parser for FOSDEM schedule data in pentabarf XML format.
 *
 * @author Christophe Beyls
 */
class EventsParser @Inject constructor(
    @Named("Conference") private val conferenceZoneId: ZoneId
) : Parser<Sequence<DetailedEvent>> {

    override fun parse(source: BufferedSource): Sequence<DetailedEvent> {
        val parser: XmlPullParser = xmlPullParserFactory.newPullParser().apply {
            setInput(source.inputStream(), null)
        }
        return sequence {
            while (!parser.isEndDocument) {
                if (parser.isStartTag("schedule")) {
                    while (!parser.isNextEndTag("schedule")) {
                        if (parser.isStartTag) {
                            when (parser.name) {
                                "day" -> {
                                    val day = Day(
                                        index = parser.getAttributeValue(null, "index")!!.toInt(),
                                        date = LocalDate.parse(
                                            parser.getAttributeValue(null, "date")
                                        )
                                    )

                                    while (!parser.isNextEndTag("day")) {
                                        if (parser.isStartTag("room")) {
                                            val roomName: String? =
                                                parser.getAttributeValue(null, "name")

                                            while (!parser.isNextEndTag("room")) {
                                                if (parser.isStartTag("event")) {
                                                    yield(parseEvent(parser, day, roomName))
                                                }
                                            }
                                        }
                                    }
                                }

                                else -> parser.skipToEndTag()
                            }
                        }
                    }
                }
                parser.next()
            }
        }
    }

    private fun parseEvent(parser: XmlPullParser, day: Day, roomName: String?): DetailedEvent {
        val id = parser.getAttributeValue(null, "id")!!.toLong()
        var startTime: Instant? = null
        var duration: String? = null
        var slug: String? = null
        var title: String? = null
        var subTitle: String? = null
        var trackName = ""
        var trackType = Track.Type.other
        var abstractText: String? = null
        var description: String? = null
        val persons = mutableListOf<Person>()
        val attachments = mutableListOf<Attachment>()
        val links = mutableListOf<Link>()

        while (!parser.isNextEndTag("event")) {
            if (parser.isStartTag) {
                when (parser.name) {
                    "start" -> {
                        val timeString = parser.nextText()
                        if (!timeString.isNullOrEmpty()) {
                            startTime = day.date
                                .atTime(getHours(timeString), getMinutes(timeString))
                                .atZone(conferenceZoneId)
                                .toInstant()
                        }
                    }
                    "duration" -> duration = parser.nextText()
                    "slug" -> slug = parser.nextText()
                    "title" -> title = parser.nextText()
                    "subtitle" -> subTitle = parser.nextText()
                    "track" -> trackName = parser.nextText()
                    "type" -> try {
                        trackType = enumValueOf(parser.nextText())
                    } catch (e: Exception) {
                        // trackType will be "other"
                    }
                    "abstract" -> abstractText = parser.nextText()
                    "description" -> description = parser.nextText()
                    "persons" -> while (!parser.isNextEndTag("persons")) {
                        if (parser.isStartTag("person")) {
                            val person = Person(
                                    id = parser.getAttributeValue(null, "id")!!.toLong(),
                                    name = parser.nextText()!!
                            )
                            persons += person
                        }
                    }
                    "attachments" -> while (!parser.isNextEndTag("attachments")) {
                        if (parser.isStartTag("attachment")) {
                            val attachmentType = parser.getAttributeValue(null, "type")
                            val attachment = Attachment(
                                eventId = id,
                                url = parser.getAttributeValue(null, "href")!!,
                                description = parser.nextText().let { attachmentDescription ->
                                    // Use the type to replace the description if absent
                                    if (attachmentDescription.isNullOrBlank() && attachmentType != null) {
                                        attachmentType.replaceFirstChar { it.uppercaseChar() }
                                    } else {
                                        attachmentDescription
                                    }
                                }
                            )
                            attachments += attachment
                        }
                    }
                    "links" -> while (!parser.isNextEndTag("links")) {
                        if (parser.isStartTag("link")) {
                            val link = Link(
                                    eventId = id,
                                    url = parser.getAttributeValue(null, "href")!!,
                                    description = parser.nextText()
                            )
                            links += link
                        }
                    }
                    else -> parser.skipToEndTag()
                }
            }
        }

        val endTime = if (startTime != null && !duration.isNullOrEmpty()) {
            startTime + Duration.ofMinutes(getHours(duration) * 60L + getMinutes(duration))
        } else null

        val event = Event(
                id = id,
                day = day,
                roomName = roomName,
                startTime = startTime,
                endTime = endTime,
                slug = slug,
                title = title,
                subTitle = subTitle,
                track = Track(name = trackName, type = trackType),
                abstractText = abstractText,
                description = description,
                personsSummary = null
        )
        val details = EventDetails(
                persons = persons,
                attachments = attachments,
                links = links
        )
        return DetailedEvent(event, details)
    }

    /**
     * Returns the hours portion of a time string in the "hh:mm" format, without allocating objects.
     *
     * @param time string in the "hh:mm" format
     * @return hours
     */
    private fun getHours(time: String): Int {
        return Character.getNumericValue(time[0]) * 10 + Character.getNumericValue(time[1])
    }

    /**
     * Returns the minutes portion of a time string in the "hh:mm" format, without allocating objects.
     *
     * @param time string in the "hh:mm" format
     * @return minutes
     */
    private fun getMinutes(time: String): Int {
        return Character.getNumericValue(time[3]) * 10 + Character.getNumericValue(time[4])
    }
}